src/nacl/bindings/crypto_generichash.py

```diff
@@ -112,3 +119,101 @@ def generichash_blake2b_salt_personal(data,
+    :type salt: bytes
+    :param person: must be at most
+                   :py:data:`.crypto_generichash_PERSONALBYTES` long:
+                   will be zero-padded if needed
+    :type person: bytes
+    :param digest_size: must be at most
+                        :py:data:`.crypto_generichash_BYTES_MAX`;
+                        the default digest size is
+                        :py:data:`.crypto_generichash_BYTES`
+    :type digest_size: int
+    :return: an initizialized state buffer
+    :rtype: bytes
+    """
+
+    _checkparams(digest_size, key, salt, person)
+
+    statebuf = ffi.new("unsigned char[]", crypto_generichash_STATEBYTES)
+    _state = ffi.cast("struct crypto_generichash_blake2b_state *", statebuf)
+
+    # both _salt and _personal must be zero-padded to the correct length
+    _salt = ffi.new("unsigned char []", crypto_generichash_SALTBYTES)
+    _person = ffi.new("unsigned char []", crypto_generichash_PERSONALBYTES)
+
+    ffi.memmove(_salt, salt, len(salt))
+    ffi.memmove(_person, person, len(person))
+
+    rc = lib.crypto_generichash_blake2b_init_salt_personal(_state,
+                                                           key, len(key),
+                                                           digest_size,
+                                                           _salt, _person)
+    ensure(rc == 0, 'Unexpected failure',
+           raising=exc.RuntimeError)
+
+    return statebuf
+
+
+def generichash_blake2b_update(statebuf, data):
+    """Update the blake2b hash state
+
+    :param statebuf: an initialized blake2b state buffer as returned from
+                     :py:func:`.crypto_generichash_blake2b_init`
+    :type name: bytes
+    :param data:
+    :type data: bytes
+    """
+
+    ensure(isinstance(data, bytes),
+           'Input data must be a bytes sequence',
+           raising=exc.TypeError)
+
+    _state = ffi.cast("struct crypto_generichash_blake2b_state *", statebuf)
+
+    rc = lib.crypto_generichash_blake2b_update(_state, data, len(data))
+    ensure(rc == 0, 'Unexpected failure',
+           raising=exc.RuntimeError)
+
+
+def generichash_blake2b_final(statebuf, digest_size):
+    """Finalize the blake2b hash state and return the digest.
+
+    :param statebuf:
+    :type statebuf: bytes
+    :param digest_size:
+    :type digest_size: int
+    :return: the blake2 digest of the passed-in data stream
+    :rtype: bytes
+    """
+
+    _digest = ffi.new("unsigned char[]", crypto_generichash_BYTES_MAX)
+    _state = ffi.cast("struct crypto_generichash_blake2b_state *", statebuf)
+    rc = lib.crypto_generichash_blake2b_final(_state, _digest, digest_size)
+
+    ensure(rc == 0, 'Unexpected failure',
+           raising=exc.RuntimeError)
+    return ffi.buffer(_digest, digest_size)[:]
+
+
+def generichash_blake2b_state_copy(statebuf):
+    """Return a copy of the given blake2b hash state"""
+
+    _statebuf = ffi.new("unsigned char[]", crypto_generichash_STATEBYTES)
```

Comment:
@reaperhulk After stumbling in a custom version of https://bitbucket.org/cffi/cffi/issues/191 I realized we are using the crypto_generichash_blake2b_state * as a opaque buffer, since we don't really get into its internal structure. With my latest commit, there are neither casts nor crashes, and there is no need for a C-side allocator/deallocator couple.  Would you mind starting another review pass?